[Woo POS] Clear order when exiting the POS

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #15180
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Before exiting point of sale, we warn the merchant about losing their in-progress order.

In this PR, we now actually clear it down, as any accidental retention can cause issues especially when reconnecting card readers.

This is due to a memory leak which still needs to be fixed. Without this mitigation, multiple payments are attempted when connecting a reader, if you’ve got any leaked POS instances in memory with unpaid orders saved in them.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Set up Proxyman or Charles to intercept traffic.

1. Launch the app – leave the card reader disconnected for now
2. Navigate to POS
3. Add an item to your cart
4. Tap `Check out`
5. Exit POS
6. Repeat steps 2-5 a few times, using different order items so the totals are different.
7. Repeat steps 2-4, but then connect a card reader, ideally a WisePad3
8. Observe that the payment is started for the expected amount – you can confirm this on the reader's screen (the iPad number being correct isn't enough.)
9. Observe that there's no unexpected "success" before tapping the card.
10. Observe that only the latest order is fetched when you connect the card reader, not the older previous orders.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

This does not fix the memory leaks – it just prevents the older orders being used for payments when you connect the card reader.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

joshheald committed:
15180 Clear order when exiting the POS

Before exiting point of sale, we warn the merchant about losing their in-progress order.

We need to actually clear it down as any accidental retention can cause issues especially when reconnecting card readers.

This is due to a memory leak which still needs to be fixed. Without this mitigation, multiple payments are attempted when connecting a reader, if you’ve got any leaked POS instances in memory with unpaid orders saved in them.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/238">21.8</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -474,6 +476,18 @@ extension PointOfSaleAggregateModel {
+        // We need to clear it down as any accidental retention can cause issues especially when reconnecting card readers.
+        orderController.clearOrder()
+        startPaymentOnCardReaderConnection?.cancel()
+        cardReaderDisconnection?.cancel()
```
Do you see value in canceling all the other `cancellables`?

joshheald replied:
I can't see a downside, and it does seem strange to just cancel some of them. Updated in 0b18d06

joshheald committed:
15180 Add test for clearing the order

joshheald committed:
Merge branch 'trunk' into issue/15180-mitigate-payment-issues-caused-by-pos-memory-leaks

joshheald committed:
15180 cancel all subscriptions when leaving POS

joshheald committed:
Fix merge issue

joshheald committed:
15180 Clean up POS in `onDisappear`

This avoids us adding other exit routes in future which don’t clean up the order.

staskus reviewed (approved):
I haven't reproduced an unexpected success views and correct orders are paid 👍
